### DISCUSSION TITLE
Issue with the `extends` priority

### DISCUSSION THREAD

**Discussion by lucasvazq** [eslint/eslint#15704] [Q&A] [answered]
I have a configuration that extends the following rules:
- `airbnb-typescript`
- `plugin:@typescript-eslint/recommended`

They both set a rule for `@typescript-eslint/no-empty-function`:
The first sets `error, {allow: ["arrowFunctions", "functions", "methods"]`.
The last sets just `error`.

I want to set a base rules from the airbnb plugin and then wrap them (overriding) with the recommended plugin. But, not matters how I order my array, the airbnb rules takes precedence over the recommended plugin rules. How can I fix this?

**Comment by chiefmikey** [comment]
If you manually include that rule with the preferred definition in your config's `rules` property object it will be used over the extensions.

```js
{
  extends: ['airbnb-typescript', 'plugin:@typescript-eslint/recommended'],
  rules: {
    '@typescript-eslint/no-empty-function': ['error'],
  },
};
```

If you don't want to detail each conflicting rule manually you could try moving just `plugin:@typescript-eslint/recommended` to an override.

```js
{
  extends: ['airbnb-typescript'],
  overrides: [
    {
      files: ['*.ts', '*.tsx'],
      extends: ['plugin:@typescript-eslint/recommended'],
    },
  ],
};
```

**Reply by lucasvazq** [reply]
Yeah, that's a good workaround, but they don't completely solve the problem..

There are two drawbacks with this solution:
- A) I need to know all the rules that conflict between the packages defined in `extends`. In the example I mention two that are very common, but they can be more and unknown.
- B) I need to use `override` with a shared settings, that can lead to another kind of problems.

Probably, I'll open an issue after some tests in a few days.

**Comment by chiefmikey** [comment, marked as answer]
Looking further into the [eslint docs](https://eslint.org/docs/user-guide/configuring/configuration-files#extending-configuration-files) it seems that the configs in the extends array actually do take precedence based on their order, however depending on the provided definition conflicting rule properties are essentially merged and not just replaced by the next extension.

For example:
extends[0]: `@typescript-eslint/no-empty-function: ['error', { allow: ['arrowFunctions', 'functions', 'methods'] }]`
extends[1]: `@typescript-eslint/no-empty-function: 'warn'`
result: `@typescript-eslint/no-empty-function: ['warn', { allow: ['arrowFunctions', 'functions', 'methods'] }]`

So that explains why using `airbnb-typescript` and then `plugin:@typescript-eslint/recommended` results in no change to the `airbnb-typescript` rule. From what I understand now using an override just appends the extends array to the root config extends so that would behave the same and wouldn't fully replace the rule properties either. 

The best solution I can think of right now is to reference all of the extended config rule objects then combine them into one object to use in your config. Using spread operators will overwrite any conflicting properties with the one that was included last.

```js
const rules1 = require(path.resolve(
  path.dirname(require.resolve('eslint-config-airbnb-typescript')),
  'lib/shared',
)).rules;

const rules2 = require(path.resolve(
  path.dirname(require.resolve('eslint-config-airbnb-typescript')),
  'index',
)).rules;

const rules3 = require(path.resolve(
  path.dirname(require.resolve('@typescript-eslint/eslint-plugin')),
  'configs/recommended',
)).rules;

module.exports = {
  rules: {
    ...rules1,
    ...rules2,
    ...rules3,
  },
};
```

**Reply by lucasvazq** [reply]
This is great! First of all, thank you very much for taking the time to respond. I think this solution is correct. Although it is not the way I expect to implement my eslint config, it works for any case without having to know the rules that generate conflicts and it does not bring the problems that using `overrides` could bring.

**Comment by lucasvazq** [comment]
Just for reference. I leave the lines of code that generates this behavior: https://github.com/eslint/eslintrc/blame/bdce01aad747393d13cd1ae17d8280e54c9c5191/lib/config-array/config-array.js#L242:L252

They come from the second commit of the repo: https://github.com/eslint/eslintrc/commit/9fb9c5da9b9540123a93b096557e9c249a2671ef
